File: sdks/go/mailngine/emails.go
package mailngine

import (
	"context"
	"fmt"
)

// EmailsResource provides access to the email sending and retrieval API.
type EmailsResource struct {
	client *Client
}

// Send sends an email with the given parameters.
func (r *EmailsResource) Send(ctx context.Context, params *SendEmailParams) (*Email, error) {
	var email Email
	if err := r.client.do(ctx, "POST", "/v1/emails", params, &email); err != nil {
		return nil, err
	}
	return &email, nil
}

// Get retrieves a single email by its ID.
func (r *EmailsResource) Get(ctx context.Context, id string) (*Email, error) {
	var email Email
	if err := r.client.do(ctx, "GET", "/v1/emails/"+id, nil, &email); err != nil {
		return nil, err
	}
	return &email, nil
}

// List returns a paginated list of emails.
func (r *EmailsResource) List(ctx context.Context, opts *ListOptions) (*ListResponse[Email], error) {
	path := "/v1/emails"
	if opts != nil {
		path += buildListQuery(opts)
	}

	var emails []Email
	meta, err := r.client.doWithMeta(ctx, "GET", path, nil, &emails)
	if err != nil {
		return nil, err
	}

	resp := &ListResponse[Email]{Data: emails}
	if meta != nil {
		resp.Meta = *meta
	}
	return resp, nil
}

// buildListQuery encodes ListOptions into URL query parameters.
func buildListQuery(opts *ListOptions) string {
	if opts == nil {
		return ""
	}

	q := "?"
	sep := ""
	if opts.Page > 0 {
		q += fmt.Sprintf("%spage=%d", sep, opts.Page)
		sep = "&"
	}
	if opts.PerPage > 0 {
		q += fmt.Sprintf("%sper_page=%d", sep, opts.PerPage)
	}

	if q == "?" {
		return ""
	}
	return q
}

File: sdks/node/src/resources/emails.ts
import type { Mailngine } from '../client.js';
import type { Email, ListOptions, ListResponse, SendEmailParams } from '../types.js';

/**
 * Resource for sending and retrieving emails.
 *
 * @example
 * ```ts
 * const email = await client.emails.send({
 *   from: 'hello@example.com',
 *   to: ['user@example.com'],
 *   subject: 'Welcome!',
 *   html: '<h1>Hello</h1>',
 * });
 * ```
 */
export class EmailsResource {
  constructor(private readonly client: Mailngine) {}

  /**
   * Send an email.
   *
   * POST /v1/emails
   */
  async send(params: SendEmailParams): Promise<Email> {
    return this.client.request<Email>('/v1/emails', {
      method: 'POST',
      body: params,
    });
  }

  /**
   * Retrieve a single email by ID.
   *
   * GET /v1/emails/:id
   */
  async get(id: string): Promise<Email> {
    return this.client.request<Email>(`/v1/emails/${encodeURIComponent(id)}`);
  }

  /**
   * List emails with pagination.
   *
   * GET /v1/emails
   */
  async list(options?: ListOptions): Promise<ListResponse<Email>> {
    const params = new URLSearchParams();
    if (options?.page != null) params.set('page', String(options.page));
    if (options?.perPage != null) params.set('per_page', String(options.perPage));

    const qs = params.toString();
    const path = qs ? `/v1/emails?${qs}` : '/v1/emails';

    return this.client.request<ListResponse<Email>>(path, { envelope: 'list' });
  }
}
